nodestream/pipeline/pipeline.py

```diff
@@ -91,22 +198,33 @@ class Pipeline(AggregatedIntrospectiveIngestionComponent):
     def __init__(self, steps: List[Step], step_outbox_size: int) -> None:
         self.steps = steps
         self.step_outbox_size = step_outbox_size
+        self.logger = getLogger(self.__class__.__name__)
+        self.errors = []
 
     async def run(
```

Comment:
Same thing here with `run` as I said with `work_loop`. I think we can tell a higher level story and delegate to helper functions. 

```python
async def run(self, ...):
     step_tasks = self.build_steps_into_tasks()
     return_states = await asyncio.gather(*tasks, return_exceptions=True)
     self.propogate_errors_from_return_states(return_states)
```